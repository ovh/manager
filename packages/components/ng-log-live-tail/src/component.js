import template from './template.html';
import controller from './controller';

export default {
  controller,
  template,
  transclude: {
    tile: '?tile',
    description: '?description',
  },
  bindings: {
    logApiUrl: '<',
    logKind: '<',
    logKeys: '<',
  },
};
