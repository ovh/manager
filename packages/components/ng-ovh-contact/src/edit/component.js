import controller from './controller';
import template from './template.html';

export default {
  require: {
    ovhContactCtrl: '^ovhContact',
  },
  template,
  controller,
  bindings: {
    options: '=?ovhContactChoiceOptions',
    customList: '=?ovhContactChoiceCustomList',
  },
};
