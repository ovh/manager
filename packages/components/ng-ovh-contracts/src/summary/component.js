import template from './template.html';

const component = {
  bindings: {
    contracts: '<items',
    inlineContracts: '<inlineItems',
    model: '=',
    name: '@?',
  },
  template,
};

export default component;
