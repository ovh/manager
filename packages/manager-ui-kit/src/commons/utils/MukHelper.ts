export function hashCode(value: unknown): number {
  const str = (() => {
    switch (typeof value) {
      case 'number':
      case 'bigint':
      case 'boolean':
      case 'string':
        return String(value);
      case 'object':
        return value === null ? 'null' : JSON.stringify(value);
      case 'function':
        return 'function';
      case 'undefined':
        return 'undefined';
      default:
        return 'symbol';
    }
  })();

  let hash = 0;
  for (let i = 0; i < str.length; i++) {
    hash = (hash * 32 - hash + str.charCodeAt(i)) % 2147483647;
  }

  return hash;
}
