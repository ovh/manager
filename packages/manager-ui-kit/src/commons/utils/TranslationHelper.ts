import i18next from 'i18next';
import type { Resource, ResourceLanguage } from 'i18next';

import { locales } from '@/commons/settings/Locales.constants';

/**
 * Manager Fallback Language
 */
const FALLBACK_LANGUAGE = 'fr_FR';

/**
 * Normalizes a language code to a format compatible with i18next loaders.
 * Converts hyphenated codes to underscore format and ensures region codes exist.
 *
 * Examples:
 * - 'fr-FR' → 'fr_FR'
 * - 'en' → 'en_GB'
 * - 'de' → 'de_DE'
 *
 * @param language - The language code from browser or user input.
 * @returns Normalized language code in the format `ll_RR` (e.g., 'en_GB').
 */
export const normalizeLanguageCode = (language: string): string => {
  let normalizedLang = language.replace('-', '_');

  if (!normalizedLang.includes('_')) {
    normalizedLang =
      normalizedLang === 'en' ? 'en_GB' : `${normalizedLang}_${normalizedLang.toUpperCase()}`;
  }

  return normalizedLang;
};

/**
 * Builds a resource loader function for dynamic translation loading.
 *
 * @param translationLoaders - A map of language codes to dynamic import functions for translations.
 * @param namespace - The namespace under which translations will be added.
 */
export const buildTranslationResources =
  (
    translationLoaders: Record<
      string,
      () => Promise<{ default?: Resource; [key: string]: unknown }>
    >,
    namespace: string,
  ) =>
  async (language: string): Promise<boolean> => {
    const normalizedLang = normalizeLanguageCode(language);

    // Always load fallbackLang once, before the user language
    if (!i18next.hasResourceBundle(FALLBACK_LANGUAGE, namespace)) {
      const fallbackLoader = translationLoaders[FALLBACK_LANGUAGE];
      if (fallbackLoader) {
        try {
          console.info(
            `Loading fallback language: ${FALLBACK_LANGUAGE} for namespace: ${namespace}`,
          );
          const fallbackModule = await fallbackLoader();
          const resources = (fallbackModule.default ?? fallbackModule) as Resource;
          i18next.addResources(FALLBACK_LANGUAGE, namespace, resources);
        } catch (error) {
          console.error(`Failed to load fallback translations (${FALLBACK_LANGUAGE}):`, error);
        }
      }
    }

    // Then load the requested language
    if (
      normalizedLang !== FALLBACK_LANGUAGE &&
      !i18next.hasResourceBundle(normalizedLang, namespace)
    ) {
      const loader = translationLoaders[normalizedLang];
      if (loader) {
        try {
          const module = await loader();
          const resources = (module.default ?? module) as Resource;
          i18next.addResources(normalizedLang, namespace, resources);
        } catch (error) {
          console.warn(
            `Could not load ${normalizedLang}. Falling back to ${FALLBACK_LANGUAGE}.`,
            error,
          );
        }
      }
    }

    return true;
  };

/**
 * Initializes and manages i18next language updates and resource loading.
 * Listens to i18next events and browser language changes to update translations dynamically.
 *
 * @param translationLoaders - A map of language codes to dynamic import functions.
 * @param namespace - The namespace to be used for loading translations.
 */
export const buildTranslationManager = (
  translationLoaders: Record<string, () => Promise<{ default?: Resource; [key: string]: unknown }>>,
  namespace: string,
): void => {
  const loadTranslations = buildTranslationResources(translationLoaders, namespace);

  const handleLanguageChange = async (lang: string): Promise<void> => {
    const normalizedLang = normalizeLanguageCode(lang);
    await loadTranslations(normalizedLang);
    console.log('Language changed to:', normalizedLang);
  };

  // Ensure promises are properly handled for event-based hooks
  if (i18next.isInitialized) {
    void handleLanguageChange(i18next.language);
  } else {
    i18next.on('initialized', () => {
      void handleLanguageChange(i18next.language);
    });
  }

  i18next.on('languageChanged', (lang: string) => {
    void handleLanguageChange(lang);
  });
};

/**
 * Union type of supported locale identifiers.
 */
export type LocaleType = (typeof locales)[number];

/**
 * Asynchronous loader that returns an i18next-compatible language resource.
 */
export type TranslationLoaderType = () => Promise<ResourceLanguage>;

/**
 * Creates a resolver function that unwraps a dynamically imported module’s default export.
 *
 * @template T - Type of the resolved module content
 * @returns A function that converts `import('./file.json')` to a typed Promise<T>.
 *
 * @example
 * const resolveTranslationModule = createImportResolver<ResourceLanguage>();
 * const resource = await resolveTranslationModule(import('./Messages_fr_FR.json'));
 */
const createImportResolver =
  <T>() =>
  (p: Promise<{ default: unknown }>) =>
    p.then((m) => m.default as T);

/**
 * A specialized resolver for translation JSON modules,
 * producing i18next {@link ResourceLanguage} objects.
 */
export const resolveTranslationModule = createImportResolver<ResourceLanguage>();
