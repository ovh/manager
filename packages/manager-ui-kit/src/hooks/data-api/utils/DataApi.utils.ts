import type { ColumnSort } from '@tanstack/react-table';

import type { Filter } from '@ovh-ux/manager-core-api';

export const filterEquals = (a: Filter, b: Filter) =>
  a.key === b.key && a.value === b.value && a.comparator === b.comparator;

export const getSearchParamsObject = (search: URLSearchParams) =>
  Object.fromEntries([...search.entries()]);

export const parseSorting = (params: URLSearchParams, defaultSorting?: ColumnSort): ColumnSort => {
  const sorting: ColumnSort = {
    id: '',
    desc: false,
  };
  if (params.has('sort')) {
    const sortParam = params.get('sort');
    if (sortParam) {
      sorting.id = sortParam;
      if (params.has('sortOrder')) {
        sorting.desc = params.get('sortOrder') === 'desc';
      }
    }
  } else if (defaultSorting) {
    return defaultSorting;
  }
  return sorting;
};
