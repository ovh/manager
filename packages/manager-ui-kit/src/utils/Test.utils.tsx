import { ReactNode } from 'react';

import { ShellContext, ShellContextType } from '@ovh-ux/manager-react-shell-client';

import { render } from '../../setupTest';

export const mockTrackPage = vitest.fn();
export const mockGetEnvironment = vitest.fn();

export const shellContext = {
  shell: {
    environment: {
      getEnvironment: mockGetEnvironment,
    },
    tracking: {
      trackPage: mockTrackPage,
    },
    navigation: {
      navigateTo: vitest.fn(),
      reload: vitest.fn(),
    },
    ux: {
      hidePreloader: vitest.fn(),
    },
  },
};

export const renderWithContext = ({ children }: { children: ReactNode }) =>
  render(
    <ShellContext.Provider value={shellContext as unknown as ShellContextType}>
      {children}
    </ShellContext.Provider>,
  );
