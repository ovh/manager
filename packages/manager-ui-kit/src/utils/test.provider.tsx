import React, { ComponentType } from 'react';
import { I18nextProvider, initReactI18next } from 'react-i18next';
import i18n from 'i18next';
import { render, RenderOptions, RenderResult } from '@testing-library/react';
import { QueryClient, QueryClientProvider } from '@tanstack/react-query';
import '@testing-library/jest-dom';
import 'element-internals-polyfill';

i18n.use(initReactI18next).init({
  fallbackLng: 'fr_FR',
  interpolation: {
    escapeValue: false,
  },
  resources: {
    en: {
      'action-menu': {
        common_actions: 'Actions',
      },
    },
  },
});

export const queryClient = new QueryClient({
  defaultOptions: {
    mutations: {
      retry: false,
    },
    queries: {
      retry: false,
    },
  },
});

const Wrappers = ({ children }: { children: React.ReactElement }) => {
  return (
    <QueryClientProvider client={queryClient}>
      <I18nextProvider i18n={i18n}>{children}</I18nextProvider>
    </QueryClientProvider>
  );
};

const customRender = (
  ui: React.JSX.Element,
  options?: Omit<RenderOptions, 'queries'>,
): RenderResult =>
  render(ui, { wrapper: Wrappers as ComponentType, ...options });

export { customRender as render };
