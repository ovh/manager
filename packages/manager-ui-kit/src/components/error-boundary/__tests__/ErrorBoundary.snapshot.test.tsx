import { vitest } from 'vitest';

import { mockGetEnvironment, renderWithContext } from '../../../utils/Test.utils';
import { ErrorBoundary } from '../ErrorBoundary.component';

vitest.mock('react-router-dom', (importOriginal) => ({
  ...importOriginal(),
  useRouteError: vitest.fn(),
  useMatches: () => ({
    pathname: 'vrackServices',
  }),
}));

mockGetEnvironment.mockResolvedValue({
  applicationName: 'test-application',
});

describe('ErrorBoundary Tests', () => {
  beforeEach(() => {
    vitest.clearAllMocks();
  });

  it('should render Error Boundary component when error occurs', () => {
    const { container } = renderWithContext({
      children: <ErrorBoundary redirectionApp="test" />,
    });
    expect(container.firstChild).toMatchSnapshot();
  });
});
