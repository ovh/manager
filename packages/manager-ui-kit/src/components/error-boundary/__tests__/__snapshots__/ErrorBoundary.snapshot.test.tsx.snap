// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ErrorBoundary Tests > should render Error Boundary component when error occurs 1`] = `
<section
  class="mx-auto w-full h-full max-w-[600px] overflow-hidden mx-autogrid p-5"
>
  <img
    alt="OOPS"
    class="w-full"
    src="/public/assets/error-banner-oops.png"
  />
  <h1
    class="_text--heading-1_1aw48_13 pt-2"
    data-ods="text"
  >
     
    Oups ! Une erreur est survenue.
  </h1>
  <h2
    class="_text--heading-2_1aw48_19 pt-1 pb-2"
    data-ods="text"
  >
     
    Une erreur s'est produite, veuillez réessayer plus tard.
  </h2>
  <div
    class="overflow-hidden mt-5 py-2"
  >
    <button
      class="_button_1sbbz_2 _button--primary_1sbbz_99 _button--md_1sbbz_41 _button--ghost_1sbbz_150 error-template-actions w-full"
      data-ods="button"
      data-testid="error-template-action-home"
      type="button"
    >
      Retour à la page d'accueil
    </button>
    <button
      class="_button_1sbbz_2 _button--primary_1sbbz_99 _button--md_1sbbz_41 _button--default_1sbbz_48 error-template-actions w-full"
      data-ods="button"
      data-testid="error-template-action-reload"
      type="button"
    >
      Réessayer
    </button>
  </div>
</section>
`;
